doc/src/documentation_style_guide.rst

```diff
@@ -0,0 +1,1019 @@
+something that rendered differently in code and in math. For example, the Greek
+letter α would be written as ``alpha`` in code and $\alpha$ in LaTeX. The
+reason being that $\alpha$ cannot be used in contexts referring to Python code
+because it is not valid Python, and conversely ``alpha`` would be incorrect in
+a math context because it does not render as the Greek letter (α).
+
+Style Preferences
+-----------------
+
+**Spelling and Punctuation**
+
+All narrative writing in SymPy follows American spelling and punctuation
+standards. For example, “color” is preferred over “colour” and commas should be
+placed inside of quotation marks.
+
+**Examples**
+
+::
+
+  If the ``line_color`` aesthetic is a function of arity 1, then the coloring is
+  a function of the x value of a point.
+
+  The term "unrestricted necklace," or "bracelet," is used to imply an object
+  that can be turned over or a sequence that can be reversed.
+
+If there is any ambiguity about the spelling of a word, such as in the case of
+a function named after a person, refer to the spelling of the actual SymPy
+function.
+
+For example, Chebyshev polynomials are named after Pafnuty Lvovich Tchebychev,
+whose name is sometimes transliterated from Russian to be spelled with a “T,”
+but in SymPy it should always be spelled “Chebyshev” to refer to the SymPy
+function.
+
+**Example**
+
+::
+
+  class chebyshevt(OrthogonalPolynomial):
+      r"""
+      Chebyshev polynomial of the first kind, $T_n(x)$
+      ...
+
+      """
+
+**Capitalization**
+
+Title case capitalization is preferred in all SymPy headings.
+
+**Example**
+
+::
+
+  What is Symbolic Computation?
+  -----------------------------
+
+**Grammar**
+
+WORK IN PROGRESS: Google Season of Docs technical writer will track common
+grammar mistakes (e.g., using certain articles correctly) while editing and add
+to this guide.
+
+Tone Preferences
+----------------
+
+Across SymPy documentation, please write in:
+
+* The present tense (e.g., In the following section, we are going to learn...)
+* The first-person inclusive plural (e.g., We did this the long way, but now we
+  can try it the short way...)
+* Use the generic pronoun “you” instead of “one.” Or use “the reader” or “the
+  user.” (e.g., You can access this function by... The user can then access
+  this function by...)
+* Use the gender-neutral pronoun “they” instead of “he” or “she.” (e.g., A good
+  docstring tells the user exactly what they need to know.)
+
+Avoid extraneous or belittling words such as “obviously,” “easily,” “simply,”
+“just,” or “straightforward.”
+
+Avoid unwelcoming or judgement-based phrases like “That is wrong.” Instead use
+friendly and inclusive language like “A common mistake is...”
+
+Avoid extraneous phrases like, “we just have to do one more thing.”
+
+Docstring Guidelines
+====================
+
+To contribute to SymPy’s docstrings, please read these guidelines in full.
```

Comment:
SymPy have quite a few module level docstrings. I think we either need to decide not to allow module level docstrings (these can often be moved to a narrative documentation peice) or we need some guidelines on writing module level docstrings. numpydoc does not have recommendations for this (I think).

Here is an example that comes to mind: https://github.com/sympy/sympy/blob/master/sympy/utilities/codegen.py

Reply:
Another: https://github.com/sympy/sympy/blob/master/sympy/utilities/autowrap.py

Reply:
I'm not sure why would disallow them. They show up in Sphinx just fine if you use automodule. For example, here is autowrap https://docs.sympy.org/dev/modules/utilities/autowrap.html#module-sympy.utilities.autowrap, which is pulled in from `automodule` https://raw.githubusercontent.com/sympy/sympy/master/doc/src/modules/utilities/autowrap.rst. The module docstrings also can be quite useful for people reading the source code. 

Putting some stuff here about them is a good idea though. I would say that they should give a general overview of the module. Documentation that is specific to a function or class in the module should be in the docstring of that function or class. The module level docstring should discuss the purpose and scope of the module, and give a high level example of how to use the functions/classes in the module. 

Anything else that should be mentioned about them?

Reply:
It seems to me that most of the module level docstrings I've come across seem more appropriate for a narrative page that describes a particular module or package. It also means that this is now mixed up in the API documentation, which these module level docstrings don't seem to be API specs like the class and function docstrings are. I'd like to see this stuff in the narrative, not the API. But, your module may have some API details: module level constants, import conventions, ... (not really sure).

> The module level docstring should discuss the purpose and scope of the module, and give a high level example of how to use the functions/classes in the module.

I think that would be better in a narrative documentation section, personally. It seems out of place in the API docs.

If we allow them, then I think we should have some guidelines: what should and shouldn't be in that section, what should the format be (headers, types of headers, etc).

I'd like to see us have a clear delineation in the narrative and API docs and what should go in each.

Reply:
Good to know. I'm fine having them as long as we have a guide about what is appropriate to be in them. I am also a fan of separating the narrative style docs from the API docs. For example, I think Pandas does this well:

API example: https://pandas.pydata.org/pandas-docs/version/0.25/reference/io.html

Narrative example: https://pandas.pydata.org/pandas-docs/version/0.25/user_guide/text.html

Reply:
Another reason to have module level docstrings: in Sphinx, when you add `.. automodule:: sympy.submod`, it puts the module level docstring there. You generally want to do that at the top of the file that has the API reference for that module. I'm discovering in https://github.com/sympy/sympy/pull/17698 that a lot of places do things like <code>:mod:\`sympy.submod\`</code>, but unless there is an `automodule` for `sympy.submod` somewhere, that won't be able to link to anything (and after my PR is merged it will give an error). So you always want to have an `automodule` for every module that is documented, and it might as well have at least a one line docstring to include there.

Reply:
Actually I guess `automodule` pulls in the docstring but if you just use `module` it creates a reference target but doesn't pull in the docstring.